contrib/src/main/java/com/datatorrent/contrib/hbase/AbstractHBaseWindowAppendOutputOperator.java

```diff
@@ -46,8 +40,6 @@
  * for the tuple in the table.<br>
  *
  * <br>
- * This class provides batch append where tuples are collected till the end
- * window and they are appended on end window
```

Comment:
I was also thinking earlier they look similar and couldn't figure out a better way to avoid this. Even though the concrete implementations look similar they will function differently because of the operators they extend. Also, with the changes, both will support multiple tables.

The difference between the non-window and the window operator is this. In the case of restart from failure, the non-window operator writes the tuples from all the windows replayed from the checkpoint. The windowed operator doesn't write tuples belonging to those windows after the checkpoint that were completely processed in the prior run, before the operator failed. It only repeats the partial writes for the window during which the operator failed. This reduces the number of duplicates in cases where the writes will result in new entries in HBase. Exactly once is not being done because transaction support is not available in HBase. This functionality requires the user to provide additional resources from hbase such as a meta column family etc, which is why I think there were two operators created and not a single one.

In either case when tuples are written, they are not flushed immediately to hbase server for performance reasons. They are given to hbase api which would internally flush it when some internal conditions are satisfied like the buffer filling up. However, we do want to make sure all pending tuples are flushed in certain conditions to ensure no data loss in case of operator failure. The non-window operator explicitly flushes before checkpoint and the window operator flushes during end of window because fully processed windows will be skipped on replay after failure.

In the earlier implementation of windowed operator, I noticed that the operator was batching all the tuples within the operator into one big batch and then sending all the tuples to the HBase api at the end of the window. I think this is not necessary as the HBase api internally would batch and flush appropriately. Batching at the operator level on a per window basis will introduce latency and will also significantly increase memory usage. That is why I changed it and one of the reasons now the concrete operators look similar.